process GATK4_MARKDUPLICATES {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine in ['singularity', 'apptainer'] && !task.ext.singularity_pull_docker_container
        ? 'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/ce/ced519873646379e287bc28738bdf88e975edd39a92e7bc6a34bccd37153d9d0/data'
        : 'community.wave.seqera.io/library/gatk4_gcnvkernel:edb12e4f0bf02cd3'}"

    input:
    tuple val(meta), path(bam)
    tuple val(meta2), path(fasta), path(fai), path(dict)

    output:
    tuple val(meta), path("${prefix}"),     emit: output
    tuple val(meta), path("${prefix}.bai"), emit: bam_index, optional: true
    tuple val(meta), path("*.metrics"),     emit: metrics,   optional: true
    tuple val("${task.process}"), val('gatk4'), eval("gatk --version 2>&1 | grep 'The Genome Analysis Toolkit' | cut -d' ' -f6"), topic: versions

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}.marddupped.bam"
    def input_list = bam.collect { "--INPUT ${it}" }.join(' ')

    def avail_mem = 3072
    if (!task.memory) {
        log.info('[GATK MarkDuplicates] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.')
    }
    else {
        avail_mem = (task.memory.mega * 0.8).intValue()
    }
    """
    gatk --java-options "-Xmx${avail_mem}M -XX:-UsePerfData" \\
        MarkDuplicates \\
        ${input_list} \\
        --OUTPUT ${prefix} \\
        --REFERENCE_SEQUENCE ${fasta} \\
        --METRICS_FILE ${prefix}.metrics \\
        --TMP_DIR . \\
        ${args}
    """
}
